File: articles/molecular-dynamics-chemical-simulation/fortran/molecular_dynamics_kernel.f90
program molecular_dynamics_kernel
  implicit none

  real(8) :: dt, r, v, a, r_new, v_new
  real(8) :: ratio, lj_energy, diffusion_estimate

  dt = 0.5d0
  r = 0.0d0
  v = 0.05d0
  a = 0.10d0

  r_new = r + v * dt + 0.5d0 * a * dt ** 2
  v_new = v + a * dt

  ratio = 1.0d0 / 1.12d0
  lj_energy = 4.0d0 * 1.0d0 * (ratio ** 12 - ratio ** 6)

  diffusion_estimate = 4.21d0 / (6.0d0 * 7.0d0)

  print *, "new_position=", r_new
  print *, "new_velocity=", v_new
  print *, "lj_energy=", lj_energy
  print *, "diffusion_estimate=", diffusion_estimate

end program molecular_dynamics_kernel

File: articles/molecular-dynamics-chemical-simulation/julia/molecular_dynamics_kernel.jl
# Molecular dynamics numerical kernel in Julia.

function lennard_jones(r, epsilon, sigma)
    ratio = sigma / r
    return 4.0 * epsilon * (ratio^12 - ratio^6)
end

function velocity_verlet_position(r, v, a, dt)
    return r + v * dt + 0.5 * a * dt^2
end

function velocity_update(v, a, dt)
    return v + a * dt
end

function diffusion_from_msd(msd, t)
    return msd / (6.0 * t)
end

r_new = velocity_verlet_position(0.0, 0.05, 0.10, 0.5)
v_new = velocity_update(0.05, 0.10, 0.5)

println("new_position=", round(r_new, digits=6))
println("new_velocity=", round(v_new, digits=6))
println("lj_energy=", round(lennard_jones(1.12, 1.0, 1.0), digits=6))
println("diffusion_estimate=", round(diffusion_from_msd(4.21, 7.0), digits=6))
